src/wasm/wat-parser-internal.h

```diff
@@ -0,0 +1,521 @@
+    }
+    return ctx.lexed();
+  }
+
+  return {};
+}
+
+std::optional<LexResult> spacechar(std::string_view in) {
+  LexCtx ctx(in);
+  ctx.take_prefix(" "sv) || ctx.take_prefix("\n"sv) ||
+    ctx.take_prefix("\r"sv) || ctx.take_prefix("\t"sv);
+  return ctx.lexed();
+}
+
+// space  ::= (' ' | format | comment)*
+// format ::= '\t' | '\n' | '\r'
+std::optional<LexResult> space(std::string_view in) {
+  LexCtx ctx(in);
+  while (ctx.size()) {
+    if (auto lexed = spacechar(ctx.next())) {
+      ctx.take(*lexed);
+    } else if (auto lexed = comment(ctx.next())) {
+      ctx.take(*lexed);
+    } else {
+      break;
+    }
+  }
+  return ctx.lexed();
+}
+
+bool LexCtx::can_finish() const {
+  // Logically we want to check for eof, parens, and space. But we don't
+  // actually want to parse more than a couple characters of space, so check for
+  // individual space chars or comment starts instead.
+  return empty() || lparen(next()) || rparen(next()) || spacechar(next()) ||
+         starts_with(";;"sv);
+}
+
+// num   ::= d:digit => d
+//         |  n:num '_'? d:digit => 10*n + d
+// digit ::= '0' => 0 | ... | '9' => 9
+std::optional<LexIntResult> num(std::string_view in) {
+  LexIntCtx ctx(in);
+  if (!ctx.take_digit()) {
+    return {};
+  }
+  while (true) {
+    bool under = ctx.take_prefix("_"sv);
+    if (!ctx.take_digit()) {
+      if (!under) {
+        return ctx.lexed();
+      }
+      return {};
+    }
+  }
+}
+
+// hexnum   ::= h:hexdigit => h
+//            | n:hexnum '_'? h:hexdigit => 16*n + h
+// hexdigit ::= d:digit => d
+//            | 'A' => 10 | ... | 'F' => 15
+//            | 'a' => 10 | ... | 'f' => 15
+std::optional<LexIntResult> hexnum(std::string_view in) {
+  LexIntCtx ctx(in);
+  if (!ctx.take_hexdigit()) {
+    return {};
+  }
+  while (true) {
+    bool under = ctx.take_prefix("_"sv);
+    if (!ctx.take_hexdigit()) {
+      if (!under) {
+        return ctx.lexed();
+      }
+      return {};
+    }
+  }
+}
+
+// uN ::= n:num         => n (if n < 2^N)
+//      | '0x' n:hexnum => n (if n < 2^N)
+// sN ::= s:sign n:num         => [s]n (if -2^(N-1) <= [s]n < 2^(N-1))
+//      | s:sign '0x' n:hexnum => [s]n (if -2^(N-1) <= [s]n < 2^(N-1))
+// sign ::= {} => + | '+' => + | '-' => -
+//
+// Note: Defer bounds and sign checking until we know what kind of integer we
+// expect.
+std::optional<LexIntResult> integer(std::string_view in) {
+  LexIntCtx ctx(in);
+  ctx.take_sign();
+  if (ctx.take_prefix("0x"sv)) {
+    if (auto lexed = hexnum(ctx.next())) {
+      ctx.take(*lexed);
+      if (ctx.can_finish()) {
+        return ctx.lexed();
+      }
+    }
+    // TODO: Add error production for unrecognized hexnum.
+    return {};
+  }
+  if (auto lexed = num(ctx.next())) {
+    ctx.take(*lexed);
+    if (ctx.can_finish()) {
+      return ctx.lexed();
+    }
+  }
+  return {};
+}
+
+// ======
+// Tokens
+// ======
+
+struct LParenTok {
+  friend std::ostream& operator<<(std::ostream& os, const LParenTok&) {
```

Comment:
Yes, these individual `<<` are used by `Token`'s `<<`, which is in turn used by googletest when one of the token equality assertions fails (although clang-tidy can't figure that out for some reason). Using `friend` is necessary for `Token`'s implementation of `<<` since without it, googletest can't find the implementation due to some arcane details of C++ name resolution I don't understand. This is semi-documented at https://google.github.io/googletest/advanced.html#teaching-googletest-how-to-print-your-values. For all the `*Tok` classes, I used `friend` like this for consistency with `Token`.